Sources/ComposableArchitecture/SharedState/PersistenceKey/AppStorageKey.swift

```diff
@@ -290,42 +290,77 @@ extension AppStorageKey: PersistenceKey {
     didSet: @escaping @Sendable (_ newValue: Value?) -> Void
   ) -> Shared<Value>.Subscription {
     let previousValue = LockIsolated(initialValue)
-    let userDefaultsDidChange = NotificationCenter.default.addObserver(
-      forName: UserDefaults.didChangeNotification,
-      object: self.store.wrappedValue,
-      queue: nil
-    ) { _ in
-      let newValue = load(initialValue: initialValue)
-      defer { previousValue.withValue { $0 = newValue } }
-      guard
-        !(_isEqual(newValue as Any, previousValue.value as Any) ?? false)
-          || (_isEqual(newValue as Any, initialValue as Any) ?? true)
-      else {
-        return
+    let removeObserver: () -> Void
+    if key.hasPrefix("@") || key.contains(".") {
+      let userDefaultsDidChange = NotificationCenter.default.addObserver(
+        forName: UserDefaults.didChangeNotification,
+        object: store.wrappedValue,
+        queue: nil
+      ) { _ in
+        let newValue = load(initialValue: initialValue)
+        defer { previousValue.withValue { $0 = newValue } }
+        func isEqual<T>(_ lhs: T, _ rhs: T) -> Bool? {
+          func open<U: Equatable>(_ lhs: U) -> Bool {
+            lhs == rhs as? U
+          }
+          guard let lhs = lhs as? any Equatable else { return nil }
+          return open(lhs)
+        }
+        guard
+          !(isEqual(newValue, previousValue.value) ?? false)
+            || (isEqual(newValue, initialValue) ?? true)
+        else {
+          return
+        }
+        guard !SharedAppStorageLocals.isSetting
+        else { return }
+        didSet(newValue)
       }
-      guard !SharedAppStorageLocals.isSetting
-      else { return }
-      didSet(newValue)
+      removeObserver = { NotificationCenter.default.removeObserver(userDefaultsDidChange) }
+    } else {
+      let observer = Observer {
+        guard !SharedAppStorageLocals.isSetting
+        else { return }
+        didSet(load(initialValue: initialValue))
+      }
+      store.wrappedValue.addObserver(observer, forKeyPath: key, options: .new, context: nil)
+      removeObserver = { store.wrappedValue.removeObserver(observer, forKeyPath: key) }
     }
     let willEnterForeground: (any NSObjectProtocol)?
     if let willEnterForegroundNotificationName {
       willEnterForeground = NotificationCenter.default.addObserver(
         forName: willEnterForegroundNotificationName,
         object: nil,
-        queue: nil
+        queue: .main
```

Comment:
I _believe_ this forces a thread hop, which means we should avoid the crash even on invalid KVO keys, but we could also do an explicit dispatch if we're not confident here.